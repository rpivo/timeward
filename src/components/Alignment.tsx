import React from 'react';
import StyledAlignment from '@styles/components/Alignment.styled';

type AlignmentProps = {
  readonly children: React.ReactNode;
  readonly vertical?: boolean;
  readonly horizontal?: boolean;
};

const Alignment: React.FC<AlignmentProps> =
  ({ children, vertical, horizontal }: AlignmentProps): JSX.Element =>
    <StyledAlignment vertical={ vertical } horizontal={ horizontal }>
      <div>{ children }</div>
    </StyledAlignment>;

export default Alignment;
