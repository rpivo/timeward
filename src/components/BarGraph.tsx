import React, { useContext } from 'react';
import { DashboardContext } from '@pages/Dashboard';
import StyledBarGraph from '@styles/components/BarGraph.styled';

const Bar: React.FC = (): JSX.Element => {
  const { store, totalSeconds } = useContext(DashboardContext);
  const heights: number[] = [];

  const getPreviousHeights = (index: number): number =>
    heights
      .slice(0, index)
      .reduce((acc, curr) => acc + curr, 0);

  return (
    <svg width='65'>
      {store.map((record, index, arr) => {
        const height = ~~((record.seconds / totalSeconds) * 100);
        heights.push(height);
        if (record.seconds) return (
          <rect
            id={`bar-monday-${index - 1}`}
            fill={record.color}
            key={index}
            height={arr[index + 1] ? `${height}%` : `${100 - getPreviousHeights(index)}%`}
            width='100%'
            x='0'
            y={`${getPreviousHeights(index)}%`}
          />
        );
        return null;
      }
      )}
    </svg>
  );
};

const BarGraph: React.FC = (): JSX.Element =>
  <StyledBarGraph>
    <Bar />
    <Bar />
    <Bar />
    <Bar />
    <Bar />
    <Bar />
    <Bar />
  </StyledBarGraph>;

export default BarGraph;
