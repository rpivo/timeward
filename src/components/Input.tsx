import React, { forwardRef } from 'react';

export type InputProps = {
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  password?: boolean;
  placeholder?: string;
  value?: string;
};

const Input = forwardRef(
  ({ onChange, password, placeholder, value }: InputProps, ref?: React.Ref<HTMLInputElement>) => {
    Input.displayName = 'Input';
    return (
      <input
        onChange={onChange}
        autoComplete={password ? 'on' : undefined}
        placeholder={placeholder}
        type={password ? 'password' : 'text'}
        ref={ref}
        value={value}
      />
    );
  },
);


export default Input;
