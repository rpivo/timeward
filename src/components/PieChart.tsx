import React, { useContext } from 'react';
import { DashboardContext } from '@pages/Dashboard';
import StyledPieChart from '@styles/components/PieChart.styled';

const PieChart: React.FC = (): JSX.Element => {
  const { store, totalSeconds } = useContext(DashboardContext);
  const strokes: number[] = [];

  const getPreviousStrokeLengths = (index: number): number =>
    strokes
      .slice(0, index)
      .reduce((acc, curr) => acc + curr, 0);

  const getRotation = (index: number): number =>
    ((getPreviousStrokeLengths(index) / 157) * 360) - 90;

  const getRandomHexColor = (): string => `#${(Math.random() * 0xFFFFFF << 0).toString(16)}`;

  return (
    <StyledPieChart>
      <svg
        width='100'
        height='100'
        viewBox='0 0 100 100'
      >
      </svg>
      {store.map((record, index) => {
        const stroke = ~~((record.seconds / totalSeconds) * 157);
        strokes.push(stroke);
        if (record.seconds) return (
          <svg
            key={index}
            width='100'
            height='100'
            viewBox='0 0 100 100'
            style={{ transform: `rotate(${getRotation(index)}deg)` }}
          >
            <circle
              r='25'
              cx='50'
              cy='50'
              fill='transparent'
              stroke={getRandomHexColor()}
              strokeDasharray={`${stroke} 157`}
              strokeWidth='50'
            />
          </svg>
        );
        return null;
      })}
    </StyledPieChart>
  );
};

export default PieChart;

