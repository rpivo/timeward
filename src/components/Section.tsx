import React from 'react';
import StyledSection from '@styles/components/Section.styled';

type SectionProps = { children: React.ReactNode };

const Section = (props: SectionProps): JSX.Element =>
  <StyledSection>{ props.children }</StyledSection>;

export default Section;
