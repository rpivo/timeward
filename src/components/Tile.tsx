import React from 'react';
import StyledTile from '@styles/components/Tile.styled';

type TileProps = {
  children: React.ReactNode;
  width?: string;
};

const Tile: React.FC<TileProps> = ({ children, width }: TileProps): JSX.Element =>
  <StyledTile width={width}>{ children }</StyledTile>;

export default Tile;
