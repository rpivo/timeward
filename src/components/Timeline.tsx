import React from 'react';
import StyledTimeline from '@styles/components/Timeline.styled';

type TimelineProps = { children: React.ReactNode };

type DayProps = {
  day: number;
  recordCount?: number;
};

type TimelineComposition = { Day: React.FC<DayProps> };

const Day: React.FC<DayProps & {}> = ({ day, recordCount }: DayProps): JSX.Element => {
  const getDayString = (): string => {
    const date = new Date();
    date.setDate(date.getDate() - day);
    return `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()}`;
  };
  return (
    <div className='day'>
      { getDayString() }
      <br />
      { recordCount }
    </div>
  );
};

const Timeline: React.FC<TimelineProps> & TimelineComposition =
  ({ children }: TimelineProps): JSX.Element =>
    <StyledTimeline>{ children }</StyledTimeline>;

Timeline.Day = Day;

export default Timeline;
