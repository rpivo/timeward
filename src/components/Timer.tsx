import React from 'react';
import { Button } from '@components/Button';

interface TimerState { time: string }

export class Timer extends React.Component <{}, TimerState> {
  public state: TimerState = { time: '00:00:00' };

  public render(): JSX.Element {
    return (
      <div>
        { this.state.time }
        <Button kind='test' />
      </div>
    );
  }
}
