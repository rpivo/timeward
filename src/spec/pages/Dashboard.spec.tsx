import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Dashboard from '@pages/Dashboard';
import Alignment from '@components/Alignment';
import TimerButton from '@components/TimerButton';
import BarGraph from '@components/BarGraph';
import Input from '@components/Input';
import PieChart from '@components/PieChart';
import Section from '@components/Section';
import Tile from '@components/Tile';
import Timeline from '@components/Timeline';
import Timer from '@components/Timer';
import Timesheet from '@components/Timesheet';

describe('Dashboard', () => {

  afterEach(() => {
    jest.restoreAllMocks();
  });

  describe('render', () => {
    const wrapper = mount(<Dashboard />);
    const dashboard = wrapper.find(Dashboard);

    const testRenderer = renderer.create(<Dashboard />);

    it('should render correctly', () => {
      testRenderer.toJSON();
      expect(testRenderer).toMatchSnapshot();
    });

    it('should contain two Alignment components', () => {
      const children = dashboard.find(Alignment);
      expect(children.length).toEqual(2);
    });

    it('should contain one Graph component', () => {
      const children = dashboard.find(BarGraph);
      expect(children.length).toEqual(1);
    });

    it('should contain one Input component', () => {
      const children = dashboard.find(Input);
      expect(children.length).toEqual(1);
    });

    it('should contain one PieChart component', () => {
      const children = dashboard.find(PieChart);
      expect(children.length).toEqual(1);
    });

    it('should contain one Section component', () => {
      const children = dashboard.find(Section);
      expect(children.length).toEqual(1);
    });

    it('should contain three Tile components', () => {
      const children = dashboard.find(Tile);
      expect(children.length).toEqual(3);
    });

    it('should contain one Timeline component', () => {
      const children = dashboard.find(Timeline);
      expect(children.length).toEqual(1);
    });

    it('should contain one Timer component', () => {
      const children = dashboard.find(Timer);
      expect(children.length).toEqual(1);
    });

    it('should contain one Timesheet component', () => {
      const children = dashboard.find(Timesheet);
      expect(children.length).toEqual(1);
    });
  });

  describe('timesheet', () => {
    describe('input', () => {
      it('should initially be set to an empty string', () => {
        const wrapper = mount(<Dashboard />);
        expect(wrapper.find(Input).find('input').getDOMNode<HTMLInputElement>().value).toBe('');
      });
    });

    describe('records', () => {
      it('should initially have 0 records', () => {
        const wrapper = mount(<Dashboard />);
        expect(wrapper.find(Timesheet).find(Timesheet.Record).length).toBe(0);
      });

      it('should have 1 record when the Timer stop Button is clicked', () => {
        const wrapper = mount(<Dashboard />);
        act(() => wrapper.find(Timer).find(TimerButton).at(1).props().handleClick());
        wrapper.update();
        expect(wrapper.find(Timesheet).find(Timesheet.Record).length).toBe(1);
      });

      it('should stack records when timer is stopped and new record name already exists', () => {
        const wrapper = mount(<Dashboard />);
        wrapper.find(Input).find('input').getDOMNode<HTMLInputElement>().value = 'test';
        act(() => wrapper.find(Timer).find(TimerButton).at(1).props().handleClick());
        wrapper.update();
        wrapper.find(Input).find('input').getDOMNode<HTMLInputElement>().value = 'test';
        act(() => wrapper.find(Timer).find(TimerButton).at(1).props().handleClick());
        wrapper.update();
        expect(wrapper.find(Timesheet).find(Timesheet.Record).length).toBe(1);
      });
    });
  });

});
