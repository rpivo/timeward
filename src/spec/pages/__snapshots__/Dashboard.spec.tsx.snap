// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard render should render correctly 1`] = `
<div
  className="sc-fzoLsD jIDzto"
>
  <section
    className="sc-AxhCb ZaNiK"
  >
    <div
      className="sc-AxhUy hrMuFC"
    >
      <div
        className="sc-AxjAm cEVkqc"
      >
        <div>
          <label
            htmlFor="what_are_you_doing?"
            style={
              Object {
                "display": "none",
              }
            }
          >
            What are you doing?
          </label>
          <input
            placeholder="What are you doing?"
            type="text"
          />
        </div>
      </div>
      <div
        className="sc-AxjAm cEVkqc"
      >
        <div>
          <div
            className="sc-Axmtr gKmsNw"
          >
            <span>
              00:00:00
            </span>
            <div
              className="sc-AxheI jGfEct"
              kind="start"
              onClick={[Function]}
            >
              <svg
                height="15"
                viewBox="0 0 50 50"
                width="15"
              >
                <path
                  d="M28.682,12.087a4.188,4.188,0,0,1,7.254,0L56.741,48.109a4.188,4.188,0,0,1-3.627,6.283H11.5a4.188,4.188,0,0,1-3.627-6.283Z"
                  fill="#aaa"
                  transform="translate(54.392 -7.309) rotate(90)"
                />
              </svg>
            </div>
            <div
              className="sc-AxheI jGfEct"
              kind="stop"
              onClick={[Function]}
            >
              <svg
                height="15"
                viewBox="0 0 50 50"
                width="15"
              >
                <path
                  d="M4.237,0H45.763A4.237,4.237,0,0,1,50,4.237V45.763A4.237,4.237,0,0,1,45.763,50H4.237A4.237,4.237,0,0,1,0,45.763V4.237A4.237,4.237,0,0,1,4.237,0Z"
                  fill="#aaa"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-fzozJi ddqBGC"
      >
        <div>
          Today: 
          00:00:00
        </div>
      </div>
    </div>
    <div
      className="sc-AxhUy hrMuFC"
    >
      <div
        className="sc-AxirZ cBNplJ"
      >
        <svg
          width="65"
        />
        <svg
          width="65"
        />
        <svg
          width="65"
        />
        <svg
          width="65"
        />
        <svg
          width="65"
        />
        <svg
          width="65"
        />
        <svg
          width="65"
        />
      </div>
    </div>
  </section>
  <div
    className="sc-AxiKw hIppFK"
  >
    <svg
      height="100"
      viewBox="0 0 100 100"
      width="100"
    />
  </div>
  <div
    className="sc-AxhUy xsjHy"
    width="full"
  >
    <div
      className="sc-AxgMl jpVdqI"
    >
      <div
        className="day"
      >
        5/10/2020
        <br />
        0
      </div>
      <div
        className="day"
      >
        5/9/2020
        <br />
      </div>
      <div
        className="day"
      >
        5/8/2020
        <br />
      </div>
      <div
        className="day"
      >
        5/7/2020
        <br />
      </div>
      <div
        className="day"
      >
        5/6/2020
        <br />
      </div>
      <div
        className="day"
      >
        5/5/2020
        <br />
      </div>
      <div
        className="day"
      >
        5/4/2020
        <br />
      </div>
    </div>
  </div>
</div>
`;
